man/retriangulate.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "RETRIANGULATE" "1" "10 June 2026" "" "The Regina Handbook"

.SH NAME
retriangulate \- Exhaustively search through triangulations or knot/link diagrams using local moves
.SH SYNOPSIS

\fBretriangulate\fR [ \fB-h, --height=\fIheight\fB\fR ] [ \fB-t, --threads=\fIthreads\fB\fR ] [ \fB-k, --knot\fR | \fB-3, --dim3\fR | \fB-4, --dim4\fR ] [ \fB-l, --all\fR ] [ \fB-g, --gen=\fIgeneration\fB\fR | \fB-a, --anysig\fR ] [ \fB-c, --classical\fR | \fB-V, --virtual\fR ] [ \fB--\fR ] \fB\fIsignature\fB\fR


\fBretriangulate\fR { \fB-v, --version\fR | \fB-?, --help\fR }

.SH "DESCRIPTION"
.PP
Given a 3-dimensional triangulation, 4-dimensional triangulation or
link diagram,
this utility uses local moves to exhaustively search for other
triangulations/diagrams of the same 3-manifold, 4-manifold or link
that are the same size or smaller.
Here "local moves" means Pachner moves for
triangulations, or Reidemeister moves for links.
.PP
Specifically, suppose the input triangulation or link diagram contains
\fIn\fR tetrahedra/pentachora/crossings
(for a 3-manifold, 4-manifold or link respectively).
Then this utility will
exhaustively modify the triangulation or link diagram using local moves,
without ever exceeding
\fIn\fR\~+\~\fIheight\fR
tetrahedra/pentachora/crossings in total.
Moreover, all such triangulations/diagrams are guaranteed to be found, each
once and only once (up to an appropriate notion of
combinatorial isomorphism).
.PP
Regarding what constitutes a "local move":
.PP
.TP 0.2i
\(bu
For 3-dimensional triangulations, this utility will only attempt 2-3
and 2-3 Pachner moves, never 1-4 or 4-1 moves.
.TP 0.2i
\(bu
For 4-dimensional triangulations, this utility will use all types of
Pachner moves.
.TP 0.2i
\(bu
For link diagrams, this utility will use all of the classical
(planar) Reidemeister moves.
There is also the option to use the \fIvirtual\fR
(non-planar) type\~II move, which adds two crossings by pushing
\fBany\fR two strands over one another from any direction,
and which could thereby add a handle to the surface in which the diagram
is embedded.
By default, the virtual type\~II move is considered only if the input
diagram has positive virtual genus (i.e., is not a classical diagram);
this behaviour can be modified using the options
\fB--classical\fR and \fB--virtual\fR\&.
.PP
The input is assumed to represent a 3-dimensional triangulation
unless one of the other link or dimension options is passed
(\fB--knot\fR or \fB--dim4\fR).
.PP
The program will output each triangulation or link diagram that it finds
of the same size \fIn\fR (including the original input
triangulation/diagram).  If it ever finds a
\fBsmaller\fR triangulation or diagram (thereby proving the
original to be non-minimal), it will output that smaller
triangulation/diagram and then stop immediately.  Otherwise it will continue
outputting triangulations or diagrams of size \fIn\fR
until no more can be found.  Although the program also finds larger
triangulations/diagrams as part of its exhaustive search using local moves,
these larger triangulations/diagrams (of which there are typically many)
will not be output unless the option \fB--all\fR is passed.
.PP
All triangulations or link diagrams, both input and output, are described
using \fIsignatures\fR\&.  These are short text strings
that identify a triangulation or link diagram uniquely up to
combinatorial isomorphism (this includes relabellings of
tetrahedra, pentachora or crossings; relabellings of individual
tetrahedron or pentachoron vertices; plus various
rotation/reflection/reversal operations on triangulations or link diagrams).
.PP
By default, the program will output
\fIsecond-generation\fR signatures (as introduced
in Regina\~8.0); you can change this by passing
\fB--gen=1\fR\&.
Both second-generation and first-generation signatures are accepted as
input, regardless of what type of output you request.
For triangulations, all output signatures will be unoriented
(which means that any specific orientation that might be encoded in your
input triangulation will be forgotten).
.PP
To view signatures in Regina's graphical user interface, visit the
Composition tab in the triangulation viewer,
or the Codes tab in the knot/link viewer.
In Python you can call x.neoSig()
for a triangulation or link diagram x\&.
.sp
.RS
.B "Tip:"
Very large triangulations or links have signatures that begin with a dash
(-).  Such a signature could be mistaken for an option
when passing it on the command line.  To avoid this, you can pass the
special option \fB--\fR immediately before it: this indicates
that all command-line options have finished, and whatever comes next should
be treated as the input signature.

.nf
    example$ \fBretriangulate -h1 -- -b-LLvALwvM...\fR
.fi
.RE
.SH "OPTIONS"
.TP
\fB-h, --height=\fIheight\fB\fR
Specifies the number of additional tetrahedra, pentachora or
crossings (for a 3-manifold, 4-manifold or link respectively)
that we allow during intermediate stages of retriangulation.
That is, if the input triangulation or link diagram
has \fIn\fR tetrahedra/pentachora/crossings, then
this utility will exhaustively search through all triangulations
or link diagrams that it can reach via local moves that do not exceed
\fIn\fR\~+\~\fIheight\fR
tetrahedra/pentachora/crossings in total.

Note that these larger intermediate triangulations or diagrams will not
be written to output; however, a larger height may allow the program
to access additional \fBsmaller\fR triangulations
or diagrams that were otherwise inaccessible.

The given height must be a non-negative integer.
In addition, for 3-dimensional triangulations it must be strictly
positive, and for 4-dimensional triangulations it must be even.

If not specified, this option defaults to 1 when working with
3-dimensional triangulations or link diagrams,
and it defaults to 2 when working with 4-dimensional triangulations.
.sp
.RS
.B "Warning:"
In general, the running time can grow
\fBsuperexponentially\fR with
\fIheight\fR\&.  It is strongly suggested
that you begin with --height=1
(or --height=2 in dimension\~4)
and increase it one step at a time
until the performance becomes unacceptable.
.RE
.TP
\fB-t, --threads=\fIthreads\fB\fR
Specifies the degree to which this utility uses parallel processing.
Specifically, this program will use
\fIthreads\fR simultaneous threads of execution
as it works its way through the different retriangulations or diagrams
of the input manifold or link.

This program is typically able to use parallelism effectively,
and so running with \fIk\fR threads should
approximately divide the running time by \fIk\fR\&.

If not specified, this option defaults to 1
(i.e., single-threaded processing, with no parallelism).
.TP
\fB-k, --knot\fR
Indicates that the given signature represents a knot or link.
The local moves used will be the Reidemeister moves, possibly including
the virtual type\~II move; see \fB--classical\fR and
\fB--virtual\fR for details.
.TP
\fB-3, --dim3 (default)\fR
Indicates that the given signature represents a 3-dimensional
triangulation.
The local moves used will be 2-3 and 3-2 Pachner moves.

This is the default if none of the
arguments \fB--knot\fR, \fB--dim3\fR or
\fB--dim4\fR is passed.
.TP
\fB-4, --dim4\fR
Indicates that the given signature represents a 4-dimensional
triangulation.
The local moves used will be 1-5, 2-4, 3-3, 4-2 and 5-1 Pachner moves.
.TP
\fB-l, --all\fR
Output the signatures of all triangulations or link diagrams that are
visited, including those that are larger than the original.

By default, larger triangulations or link diagrams (of which there may
be very many) are not included in the output.
.TP
\fB-g, --gen=\fIgeneration\fB\fR
Indicates whether to output first-generation signatures or
second-generation signatures.

Second-generation signatures (\fB--gen=2\fR) were introduced
in Regina\~8.0.  They are smaller than first-generation signatures,
often faster to compute, and are recommended for use in new projects.

First-generation signatures (\fB--gen=1\fR) were used in
Regina\~<=\~7.x.  Since Regina uses second-generation
signatures internally, outputting first-generation signatures will add
extra overhead to the running time.  You should only use these
if you need backward compatibility (e.g., if you need the match the
output against a list of signatures that was generated using an older
version of Regina).

The default is to output second-generation signatures.
.TP
\fB-a, --anysig\fR
A legacy alias for \fB--gen=2\fR\&.

Since second-generation signatures are now the default, this option has
no effect and should no longer be used.
.TP
\fB-c, --classical\fR
For link diagrams, only use the classical type\~I, II and III
Reidemeister moves.  This is the default behaviour when the input is a
classical (planar) link diagram.

This option can only be used with \fB--knot\fR\&.
.TP
\fB-V, --virtual\fR
For link diagrams, use the classical type\~I, II and III
Reidemeister moves as well as the virtual type\~II move.
The virtual type\~II move is like the classical type\~II move
to add two crossings, except that we are allowed to push
\fBany\fR two strands over one another from any direction;
as a consequence, this could add a handle to the surface in which the
link diagram is embedded.
This option is the default behaviour when the input is a link diagram
with positive virtual genus (i.e., a non-classical link diagram).

This option can only be used with \fB--knot\fR\&.
.TP
\fB--\fR
Indicates that all other options have finished, and whatever comes
next on the command line should be treated as the input signature.

This is useful when your signature begins with a dash, to avoid
confusing your input signature with a regular command line option.
.TP
\fB-v, --version\fR
Show which version of Regina is being used, and exit
immediately.
.TP
\fB-?, --help\fR
Display brief usage information, and exit immediately.
.SH "EXAMPLES"
.PP
The following 3-manifold triangulation is non-minimal, but it requires a bit
of work to see this:

.nf
    example$ \fBretriangulate -h2 hpyS1+lmKi0Kdb\fR
    hpyS1+lmKi0Kdb
    hhrSU+fnewBNLa
    Found 2 triangulation(s).
    example$ \fBretriangulate -h3 hpyS1+lmKi0Kdb\fR
    hpyS1+lmKi0Kdb
    hhrSU+fnewBNLa
    hNI2T+FabR0Qsa
    gxtJT7frusUe
    Triangulation is non-minimal!
    Smaller triangulation: gxtJT7frusUe
    example$
.fi
.PP
Although the program stopped as soon as it found a smaller
triangulation, this can be simplified even further:

.nf
    example$ \fBretriangulate gxtJT7frusUe\fR
    gxtJT7frusUe
    fhrK80iePeb
    Triangulation is non-minimal!
    Smaller triangulation: fhrK80iePeb
    example$
.fi
.PP
A little more probing shows this to be the cusped hyperbolic
manifold m112:

.nf
    example$ \fBcensuslookup fhrK80iePeb\fR
    fhrK80iePeb: 1 hit
        m112 : #2 -- Cusped hyperbolic census (orientable)
    
    example$
.fi
.SH "MACOS USERS"
.PP
If you downloaded a drag-and-drop app bundle, this utility is
shipped inside it.  If you dragged Regina to the main
Applications folder, you can run it as
/Applications/Regina.app/Contents/MacOS/retriangulate\&.
.SH "WINDOWS USERS"
.PP
The command-line utilities are installed beneath the
\fIProgram\~Files\fR directory; on some
machines this directory is called
\fIProgram\~Files\~(x86)\fR\&.
You can start this utility by running
c:\\Program\~Files\\Regina\\Regina\~8.0\\bin\\retriangulate.exe\&.
.SH "SEE ALSO"
.PP
regina-gui\&.
.SH "AUTHOR"
.PP
This utility was written by Benjamin Burton
<bab@maths.uq.edu.au>\&.
Many people have been involved in the development
of Regina; see the users' handbook for a full list of credits.


utils/retriangulate.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Exhaustively retriangulate a given manifold                           *
 *                                                                        *
 *  Copyright (c) 2017-2026, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  As an exception, when this program is distributed through (i) the     *
 *  App Store by Apple Inc.; (ii) the Mac App Store by Apple Inc.; or     *
 *  (iii) Google Play by Google Inc., then that store may impose any      *
 *  digital rights management, device limits and/or redistribution        *
 *  restrictions that are required by its terms of service.               *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public License     *
 *  along with this program. If not, see <https://www.gnu.org/licenses/>. *
 *                                                                        *
 **************************************************************************/

#include <mutex>
#include <thread>
#include <getopt.h>
#include "link/link.h"
#include "triangulation/dim3.h"
#include "triangulation/dim4.h"

std::mutex mutex;

constexpr int NO_HEIGHT = -1000;

// Command-line arguments:
int height = NO_HEIGHT;
long threads = 1;
enum {
    FLAVOUR_NONE = 0,
    FLAVOUR_DIM3 = 3,
    FLAVOUR_DIM4 = 4,
    FLAVOUR_KNOT = 100
} flavour = FLAVOUR_NONE;
bool showAll = false;
int sigGen = 2;
bool virtualMoves = false;
bool classicalMoves = false;

template <int dim> requires (dim == 3 || dim == 4)
void process(const regina::Triangulation<dim>& tri) {
    using Signature = regina::ByteSequence;

    size_t nSolns = 0;
    bool nonMinimal = false;
    std::string simpler;

    // Note: the code here is specifically tuned to the IsoSigBinary encoding
    // (which is used by retriangulate()), since it uses
    // IsoSigBinary::asString() when writing signatures to output.
    // If we ever change RetriangulateParams<...> to use a different internal
    // signature encoding, we will need to update this code also.

    tri.retriangulate(height, threads, nullptr /* tracker */,
        [&nSolns, &nonMinimal, &simpler, &tri](
                const Signature& sig, const regina::Triangulation<dim>& t) {
            if (t.size() > tri.size()) {
                if (showAll) {
                    std::string outputSig = (sigGen == 1 ? t.isoSig() :
                        regina::IsoSigBinary::asString<dim>(sig));

                    std::lock_guard<std::mutex> lock(mutex);
                    std::cout << sig << std::endl;
                    ++nSolns;
                }
                return false;
            }

            std::string outputSig = (sigGen == 1 ? t.isoSig() :
                regina::IsoSigBinary::asString<dim>(sig));

            std::lock_guard<std::mutex> lock(mutex);
            std::cout << outputSig << std::endl;

            if (t.size() < tri.size()) {
                nonMinimal = true;
                simpler = std::move(outputSig);
                return true;
            }

            ++nSolns;
            return false;
        });

    if (nonMinimal) {
        std::cerr << "Triangulation is non-minimal!" << std::endl;
        std::cerr << "Smaller triangulation: " << simpler << std::endl;
    } else
        std::cerr << "Found " << nSolns << " triangulation(s)." << std::endl;
}

void process(const regina::Link& link) {
    using Signature = regina::ByteSequence;

    size_t nSolns = 0;
    bool nonMinimal = false;
    std::string simpler;

    // Note: the code here is specifically tuned to the LinkSigBinary encoding
    // (which is used by rewrite() and rewriteVirtual()), since it uses
    // LinkSigBinary::asString() when writing signatures to output.
    // If we ever change RetriangulateParams<Link> to use a different internal
    // signature encoding, we will need to update this code also.

    std::function<bool(const Signature&, const regina::Link&)> action =
            [&nSolns, &nonMinimal, &simpler, &link](
            const Signature& sig, const regina::Link& k) {
        if (k.size() > link.size()) {
            if (showAll) {
                std::string outputSig = (sigGen == 1 ? k.knotSig() :
                    regina::LinkSigBinary::asString(sig));

                std::lock_guard<std::mutex> lock(mutex);
                std::cout << outputSig << std::endl;
                ++nSolns;
            }
            return false;
        }

        std::string outputSig = (sigGen == 1 ? k.knotSig() :
            regina::LinkSigBinary::asString(sig));

        std::lock_guard<std::mutex> lock(mutex);
        std::cout << outputSig << std::endl;

        if (k.size() < link.size()) {
            nonMinimal = true;
            simpler = std::move(outputSig);
            return true;
        }

        ++nSolns;
        return false;
    };

    if (virtualMoves || ! (classicalMoves || link.isClassical()))
        link.rewriteVirtual(height, threads, nullptr /* tracker */, action);
    else
        link.rewrite(height, threads, nullptr /* tracker */, action);

    if (nonMinimal) {
        std::cerr << "Link is non-minimal!" << std::endl;
        std::cerr << "Smaller link: " << simpler << std::endl;
    } else
        std::cerr << "Found " << nSolns << " link(s)." << std::endl;
}

void help() {
    std::cerr <<
R"help(Usage: retriangulate <isosig>
  -h, --height=<height>       Number of extra simplices/crossings (default = 1)
  -t, --threads=<threads>     Number of parallel threads (default = 1)
  -k, --knot                  Input is a knot/link signature
  -3, --dim3                  Input is a 3-D isomorphism signature (default)
  -4, --dim4                  Input is a 4-D isomorphism signature
  -l, --all                   Output larger triangulations/links also
  -g, --gen=<generation>      Output first-generation signatures (--gen=1) or
                              second-generation signatures (--gen=2, default)
  -a, --anysig                Legacy alias for --gen=2 (default)
  -c, --classical             Never allow virtual type II moves (default for
                              classical link diagrams)
  -V, --virtual               Always allow virtual type II moves (default for
                              virtual link diagrams)
  -v, --version               Show which version of Regina is being used
      --help                  Show this help message
)help";
}

int main(int argc, char* argv[]) {
    // Parse the command-line arguments.
    const char* shortOpt = ":h:t:g:34klacVv";
    struct option longOpt[] = {
        { "height", required_argument, nullptr, 'h' },
        { "threads", required_argument, nullptr, 't' },
        { "dim3", no_argument, nullptr, '3' },
        { "dim4", no_argument, nullptr, '4' },
        { "knot", no_argument, nullptr, 'k' },
        { "all", no_argument, nullptr, 'l' },
        { "gen", required_argument, nullptr, 'g' },
        { "anysig", no_argument, nullptr, 'a' },
        { "classical", no_argument, nullptr, 'c' },
        { "virtual", no_argument, nullptr, 'V' },
        { "version", no_argument, nullptr, 'v' },
        { "help", no_argument, nullptr, '_' },
        { nullptr, 0, nullptr, 0 }
    };

    int opt;
    char* endptr;
    while ((opt = getopt_long(argc, argv, shortOpt, longOpt, nullptr)) != -1) {
        switch (opt) {
            case 'h':
                height = strtol(optarg, &endptr, 10);
                if (*endptr != 0) {
                    std::cerr << "The height must be a non-negative "
                        "integer.\n\n";
                    help();
                    return 1;
                }
                break;
            case 't':
                threads = strtol(optarg, &endptr, 10);
                if (*endptr != 0) {
                    std::cerr << "The number of threads must be a positive "
                        "integer.\n\n";
                    help();
                    return 1;
                }
                break;
            case 'k':
                if (flavour && flavour != FLAVOUR_KNOT) {
                    std::cerr << "You cannot pass more than one of "
                        "--knot, --dim3, or --dim4.\n\n";
                    help();
                    return 1;
                }
                flavour = FLAVOUR_KNOT;
                break;
            case '3':
                if (flavour && flavour != FLAVOUR_DIM3) {
                    std::cerr << "You cannot pass more than one of "
                        "--knot, --dim3, or --dim4.\n\n";
                    help();
                    return 1;
                }
                flavour = FLAVOUR_DIM3;
                break;
            case '4':
                if (flavour && flavour != FLAVOUR_DIM4) {
                    std::cerr << "You cannot pass more than one of "
                        "--knot, --dim3, or --dim4.\n\n";
                    help();
                    return 1;
                }
                flavour = FLAVOUR_DIM4;
                break;
            case 'l':
                showAll = true;
                break;
            case 'g':
                sigGen = strtol(optarg, &endptr, 10);
                if (*endptr != 0 || ! (sigGen == 1 || sigGen == 2)) {
                    std::cerr << "The signature generation must be 1 or 2.\n\n";
                    help();
                    return 1;
                }
                break;
            case 'a':
                // Second-generation signatures are already the default.
                // sigGen = 2;
                break;
            case 'c':
                classicalMoves = true;
                break;
            case 'V':
                virtualMoves = true;
                break;
            case 'v':
                // If other arguments were passed, just silently ignore them
                // for now.  Ideally we would give an error in this scenario.
                std::cout << PACKAGE_BUILD_STRING << std::endl;
                return 0;
            case '_':
                help();
                return 0;
            case '?':
                if (optopt == 0) {
                    // Unknown long option.
                    std::cerr << "Unknown option: " << argv[optind - 1];
                } else {
                    // Unknown short option.
                    std::cerr << "Unknown option: -" <<
                        static_cast<char>(optopt);
                }
                std::cerr << "\n\n";
                help();
                return 1;
            case ':':
                // If an argument is missing, getopt_long() behaves as though
                // the short option were passed, and there is no indication
                // as to whether the user specified a long option instead.
                // For now we will just use the short option in our error
                // message, since the subsequent help text explains both the
                // short and long options anyway.
                std::cerr << "Missing argument: -" << static_cast<char>(optopt);
                std::cerr << "\n\n";
                help();
                return 1;
        }
    }

    // Set any undefined options to their defaults.
    if (! flavour)
        flavour = FLAVOUR_DIM3;

    std::string sig;
    if (optind < argc) {
        sig = argv[optind];
        if (optind + 1 < argc) {
            std::cerr << "Too many arguments.\n\n";
            help();
            return 1;
        }
    } else {
        if (flavour == FLAVOUR_KNOT)
            std::cerr << "Please give a knot/link signature.\n\n";
        else
            std::cerr << "Please give an isomorphism signature.\n\n";
        help();
        return 1;
    }

    // Set default arguments.
    if (height == NO_HEIGHT) {
        switch (flavour) {
            case FLAVOUR_DIM3: height = 1; break;
            case FLAVOUR_DIM4: height = 2; break;
            case FLAVOUR_KNOT: height = 1; break;
            default: /* should never occur */ break;
        }
    }

    // Run a sanity check on the command-line arguments.
    bool broken = false;
    if (height < 0) {
        std::cerr << "The height cannot be negative.\n";
        broken = true;
    } else if (flavour == FLAVOUR_DIM3 && height == 0) {
        std::cerr << "In three dimensions the height must be positive.\n";
        broken = true;
    } else if (flavour == FLAVOUR_DIM4 && (height % 2)) {
        std::cerr << "In four dimensions the height must be even.\n";
        broken = true;
    }
    if (threads <= 0) {
        std::cerr << "The number of threads must be positive.\n";
        broken = true;
    }
    if ((virtualMoves || classicalMoves) && flavour != FLAVOUR_KNOT) {
        std::cerr << "The options --classical and --virtual can only be used "
            "with --knot.\n";
        broken = true;
    }
    if (virtualMoves && classicalMoves) {
        std::cerr << "The options --classical and --virtual cannot be used "
            "together.\n";
        broken = true;
    }

    if (! broken) {
        switch (flavour) {
            case FLAVOUR_DIM3: {
                try {
                    process(regina::Triangulation<3>::fromSig(sig));
                } catch (const regina::InvalidArgument&) {
                    std::cerr << "I could not interpret the given "
                        "3-manifold isomorphism signature.\n";
                    broken = true;
                }
                break;
            }
            case FLAVOUR_DIM4: {
                try {
                    process(regina::Triangulation<4>::fromSig(sig));
                } catch (const regina::InvalidArgument&) {
                    std::cerr << "I could not interpret the given "
                        "4-manifold isomorphism signature.\n";
                    broken = true;
                }
                break;
            }
            case FLAVOUR_KNOT: {
                try {
                    process(regina::Link::fromSig(sig));
                } catch (const regina::InvalidArgument&) {
                    std::cerr << "I could not interpret the given "
                        "knot/link signature.\n";
                    broken = true;
                }
                break;
            }
            default: /* should never occur */
                break;
        }
    }

    if (broken) {
        std::cerr << '\n';
        help();
        return 1;
    } else {
        return 0;
    }
}

